presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddNotNullFiltersToJoinNode.java

```diff
@@ -0,0 +1,232 @@
+    }
+
+    private static class Rewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final Session session;
+        private final FunctionAndTypeManager functionAndTypeManager;
+
+        public Rewriter(FunctionAndTypeManager functionAndTypeManager, PlanNodeIdAllocator idAllocator, Session session)
+        {
+            this.functionAndTypeManager = functionAndTypeManager;
+            this.idAllocator = idAllocator;
+            this.session = session;
+        }
+
+        private PlanNode newJoinNode(RewriteContext<Void> context, JoinNode currentJoinNode, RowExpression newFilterExpression)
+        {
+            return new JoinNode(
+                    currentJoinNode.getSourceLocation(),
+                    idAllocator.getNextId(),
+                    currentJoinNode.getType(),
+                    context.rewrite(currentJoinNode.getLeft()),
+                    context.rewrite(currentJoinNode.getRight()),
+                    currentJoinNode.getCriteria(),
+                    currentJoinNode.getOutputVariables(),
+                    Optional.of(newFilterExpression),
+                    currentJoinNode.getLeftHashVariable(),
+                    currentJoinNode.getRightHashVariable(),
+                    currentJoinNode.getDistributionType(),
+                    currentJoinNode.getDynamicFilters());
+        }
+
+        private RowExpression buildNotNullRowExpression(List<VariableReferenceExpression> expressions)
+        {
+            List<CallExpression> isNotNullExpressions = expressions.stream()
+                    .map(x -> new CallExpression(x.getSourceLocation(), "not",
+                            functionAndTypeManager.lookupFunction("not", fromTypes(BOOLEAN)),
+                            BOOLEAN,
+                            singletonList(new SpecialFormExpression(x.getSourceLocation(), IS_NULL, BOOLEAN, x))))
+                    .collect(Collectors.toList());
+
+            return LogicalRowExpressions.and(isNotNullExpressions);
+        }
+
+        private RowExpression combinedFilter(RowExpression existingFilter, RowExpression additionalFilter)
+        {
+            requireNonNull(additionalFilter);
+            if (existingFilter != null) {
+                return LogicalRowExpressions.and(existingFilter, additionalFilter);
+            }
+            return additionalFilter;
+        }
+
+        @Override
+        public PlanNode visitJoin(JoinNode joinNode, RewriteContext<Void> context)
+        {
+            if (!SystemSessionProperties.isOptimizeNullsInJoinv2(session)) {
+                return joinNode;
+            }
+
+            List<JoinNode.EquiJoinClause> joinCriteria = joinNode.getCriteria();
+            RowExpression joinFilter = joinNode.getFilter().orElse(null);
+
+            final ArrayList<VariableReferenceExpression> leftExpressions = new ArrayList<>();
```

Comment:
```suggestion
            ImmutableList.Builder<VariableReferenceExpression> leftExpressions = ImmutableList.builder();
```
Presto prefer immutable collections in Guava. Similar for others.

Reply:
Sure, I will change this. I agree on the general principle of using immutable collections for any references to collections that can exist outside the scope of a method. This is not the case here however so I'm not sure if the rule is warranted ?